File: 04_Unsupervised_ML_Examples/Scripts_R/SCRIPT - Casas.R

# Análise de Correspondência Múltipla + Análise Fatorial PCA

# Curso: MBA DSA (USP ESALQ)

# Prof. Wilson Tarantin Jr.

# Instalação e carregamento dos pacotes utilizados

pacotes <- c("plotly",
             "tidyverse",
             "ggrepel",
             "knitr", 
             "kableExtra",
             "reshape2",
             "PerformanceAnalytics", 
             "psych",
             "ltm", 
             "Hmisc",
             "readxl",
             "sjPlot",
             "ade4")

if(sum(as.numeric(!pacotes %in% installed.packages())) != 0){
  instalador <- pacotes[!pacotes %in% installed.packages()]
  for(i in 1:length(instalador)) {
    install.packages(instalador, dependencies = T)
    break()}
  sapply(pacotes, require, character = T) 
} else {
  sapply(pacotes, require, character = T) 
}

# Carregamento da base de dados
casas <- read_xlsx("Preco Casas.xlsx")

# Fonte: https://www.kaggle.com/datasets/elakiricoder/jiffs-house-price-prediction-dataset

# Separação das variáveis em qualitativas e quantitativas
var_quali <- casas[,c(5,6,7,8,10,12,15)]
var_quanti <- casas[,c(1,2,3,4,9,11,13,14)]

## Nota: vamos deixar a variável "valor da casa" fora da análise por enquanto
## O objetivo é criar um ranking que capture os valores das casas

# A função para a criação da ACM pede que sejam utilizados "fatores"
var_quali <- as.data.frame(unclass(var_quali), stringsAsFactors=TRUE)

# Ajustando variáveis quantitativas que estão como textos
var_quanti$distance_to_school <- as.double(var_quanti$distance_to_school)
var_quanti$distance_to_supermarket_km <- as.double(var_quanti$distance_to_supermarket_km)
var_quanti$crime_rate_index <- as.double(var_quanti$crime_rate_index)

# Estatísticas descritivas
summary(var_quali)
summary(var_quanti)

# Iniciando a Análise de Correspondência Múltipla nas variáveis qualitativas

# Tabelas de contingência
sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$parking_space,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$front_garden,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$swimming_pool,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$wall_fence,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$water_front,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

sjt.xtab(var.row = var_quali$large_living_room,
         var.col = var_quali$room_size_class,
         show.exp = TRUE,
         show.row.prc = TRUE,
         show.col.prc = TRUE, 
         encoding = "UTF-8")

# Análise de Correspondência Múltipla
ACM <- dudi.acm(var_quali, scannf = FALSE)

# Analisando as variâncias de cada dimensão
perc_variancia <- (ACM$eig / sum(ACM$eig)) * 100
perc_variancia

# Quantidade de categorias por variável
quant_categorias <- apply(var_quali,
                          MARGIN =  2,
                          FUN = function(x) nlevels(as.factor(x)))

# Consolidando as coordenadas-padrão obtidas por meio da matriz binária
df_ACM <- data.frame(ACM$c1, Variável = rep(names(quant_categorias),
                                            quant_categorias))

# Plotando o mapa perceptual
df_ACM %>%
  rownames_to_column() %>%
  rename(Categoria = 1) %>%
  ggplot(aes(x = CS1, y = CS2, label = Categoria, color = Variável)) +
  geom_point() +
  geom_label_repel() +
  geom_vline(aes(xintercept = 0), linetype = "longdash", color = "grey48") +
  geom_hline(aes(yintercept = 0), linetype = "longdash", color = "grey48") +
  labs(x = paste("Dimensão 1:", paste0(round(perc_variancia[1], 2), "%")),
       y = paste("Dimensão 2:", paste0(round(perc_variancia[2], 2), "%"))) +
  theme_bw()

# Obtendo as coordenadas das observações
coord_obs <- ACM$li

# Adicionando as coordenadas ao banco de dados de variáveis quantitativas
var_quanti <- bind_cols(var_quanti, coord_obs)

# Coeficientes de correlação de Pearson para cada par de variáveis
rho <- rcorr(as.matrix(var_quanti), type="pearson")

corr_coef <- rho$r # Matriz de correlações
corr_sig <- round(rho$P, 5) # Matriz com p-valor dos coeficientes

# Elaboração de um mapa de calor das correlações de Pearson entre as variáveis
ggplotly(
    var_quanti %>%
    cor() %>%
    melt() %>%
    rename(Correlação = value) %>%
    ggplot() +
    geom_tile(aes(x = Var1, y = Var2, fill = Correlação)) +
    geom_text(aes(x = Var1, y = Var2, label = format(round(Correlação,3))),
              size = 3) +
    scale_fill_viridis_b() +
    labs(x = NULL, y = NULL) +
    theme_bw(base_size = 6))

### Elaboração a Análise Fatorial Por Componentes Principais ###

# Teste de esfericidade de Bartlett
cortest.bartlett(var_quanti)

# Elaboração da análise fatorial por componentes principais
fatorial <- principal(var_quanti,
                      nfactors = length(var_quanti),
                      rotate = "none",
                      scores = TRUE)
fatorial

# Eigenvalues (autovalores)
eigenvalues <- round(fatorial$values, 5)
eigenvalues
round(sum(eigenvalues), 2) # soma dos autovalores

# Identificação da variância compartilhada em cada fator
variancia_compartilhada <- as.data.frame(fatorial$Vaccounted) %>% 
  slice(1:3)

rownames(variancia_compartilhada) <- c("Autovalores",
                                       "Prop. da Variância",
                                       "Prop. da Variância Acumulada")

# Variância compartilhada pelas variáveis originais para a formação de cada fator
round(variancia_compartilhada, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

### Elaboração da Análise Fatorial por Componentes Principais ###
### Fatores extraídos a partir de autovalores maiores que 1 ###

# Definição da quantidade de fatores com eigenvalues maiores que 1
k <- sum(eigenvalues > 1)
print(k)

# Elaboração da análise fatorial por componentes principais
fatorial_final <- principal(var_quanti,
                            nfactors = k,
                            rotate = "none",
                            scores = TRUE)

# Cálculo dos scores fatoriais
scores_fatoriais <- as.data.frame(fatorial_final$weights)

# Visualização dos scores fatoriais
round(scores_fatoriais, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

# Cálculo dos fatores propriamente ditos
fatores <- as.data.frame(fatorial_final$scores)
View(fatores)

# Cálculo das cargas fatoriais
cargas_fatoriais <- as.data.frame(unclass(fatorial_final$loadings))

# Visualização das cargas fatoriais
round(cargas_fatoriais, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped", 
                full_width = FALSE, 
                font_size = 20)

# Cálculo das comunalidades
comunalidades <- as.data.frame(unclass(fatorial_final$communality)) %>%
  rename(comunalidades = 1)

# Visualização das comunalidades para os 2 fatores extraídos
round(comunalidades, 3) %>%
  kable() %>%
  kable_styling(bootstrap_options = "striped",
                full_width = FALSE,
                font_size = 20)

# Loading plot com as cargas dos 2 primeiros fatores
cargas_fatoriais[, 1:2] %>%
  data.frame() %>%
  rownames_to_column("variáveis") %>%
  ggplot(aes(x = PC1, y = PC2, label = variáveis)) +
  geom_point(color = "darkorchid",
             size = 3) +
  geom_text_repel() +
  geom_vline(aes(xintercept = 0), linetype = "dashed", color = "orange") +
  geom_hline(aes(yintercept = 0), linetype = "dashed", color = "orange") +
  expand_limits(x= c(-1.25, 0.25), y=c(-0.25, 1)) +
  theme_bw()

# Criação de um ranking Critério da soma ponderada e ordenamento)
casas$ranking <- fatores$PC1 * variancia_compartilhada$PC1[2] +
                 fatores$PC2 * variancia_compartilhada$PC2[2] +
                 fatores$PC3 * variancia_compartilhada$PC3[2] + 
                 fatores$PC4 * variancia_compartilhada$PC4[2]

# Ranking e valor
corr_valor <- rcorr(as.matrix(casas[,16:17]))

valor_corr_coef <- corr_valor$r # Matriz de correlações
valor_corr_sig <- round(corr_valor$P, 5) # Matriz com p-valor dos coeficientes

# Fim!

File: 04_Unsupervised_ML_Examples/SCRIPT - Casas.py
# Geral
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# PCA
from scipy.stats import pearsonr
import scipy.linalg as sla
from scipy.stats import zscore
import prince

# ANACOR Múltipla
import prince
from scipy.stats import chi2_contingency


# PCA
# 01º Passo: Matriz de correlações de Pearson
# 02° Passo: Teste de esferecidade de Barllet
# 03° Passo: Autovalores e autovetores da matriz de correlação de Pearson. Da pra tirar a variância compartilhada, que é a porcentagem de cada autovalor da soma deles
# 04° Passo: Scores fatoriais (autovetores dividido pela raiz do seu autovalor)
# 05° Passo: Zscores (são as observações padronizadas)
# 06° Passo: Fatores (multiplicação matricial dos Zscores com os scores fatoriais)
# 07° Passo: Cargas fatoriais (correlaçaõ de Pearson entre os fatores e as variáveis originais)
# 08° Passo: Critério de Kaiser (autovalores maiores que 1), para definir o número de fatores
# 09° Passo: Comunalidades dos autovalores de Kaiser
# 10° Passo: Plotar as cargas fatoriais

# ANACOR Múltipla
# 01° Passo: Tabelas de contingência
# 02° Passo: Número de dimensões (número de categorias - número de variáveis)
# 03° Passo: ANACOR pelo Prince
# 04° Passo: Mapa perceptual



#%%
# Carregamento da base de dados (Preco Casas)
df = pd.read_excel('datasets/Preco Casas.xlsx') 

#%%
# Visualização da base de dados
print('################### DATAFRAME ####################')
print(df)

#%%
# Tipos das variáveis
print('################### DATAFRAME TYPES ####################')
print(df.dtypes)

#%%
# Algumas variáveis são qualitativas e outras são quantitativas
# Vamos separar o banco de dados em 2 partes (somente quali e quanti)

var_quali = df[['large_living_room', 'parking_space', 'front_garden', 'swimming_pool', 'wall_fence', 'water_front', 'room_size_class']]
var_quanti = df[['land_size_sqm', 'house_size_sqm', 'no_of_rooms', 'no_of_bathrooms', 'distance_to_school', 'house_age', 'distance_to_supermarket_km', 'crime_rate_index']]

## Nota: vamos deixar a variável "valor da casa" fora da análise por enquanto
## O objetivo é criar um ranking que capture os valores das casas
#%%
print('############ ESTATÍSTICAS DESCRITIVAS QUALI ############')
print(var_quali.describe())

print('############ ESTATÍSTICAS DESCRITIVAS QUANTI ############')
print(var_quanti.describe())
#%%

# INÍCIO ANACOR PARA VARIÁVEIS QUALITATIVAS

#%%
# 01° Passo: Tabelas de contingência
def contingencia_chi(df, column1, column2):
    # Tabela de contingência com frequências absolutas observadas

    tabela_contingencia = pd.crosstab(index=df[column1], columns=df[column2])
    print('####### CONTINGENCY TABLE OBSERVED ##########')
    print(tabela_contingencia)

    # Estatística chi quadrado e teste
    # Se o p valor é menor que 0.05 então se rejeita  hipótese nula, isto é, que as variáveis não tem associação estatisticamente relevantes.
    chi2 = chi2_contingency(tabela_contingencia)
    stat, pvalue, dof, expected = chi2 # Estatística do teste, p valor, degrees of freedom, tabela de frequências esperadas

    print('####### CHI SQUARED TEST ##########')
    teste_chi2 = pd.DataFrame({'stat': [stat], 'pvalues':[pvalue], 'dof': [dof]}, index = ['values'])
    print(teste_chi2)
#%%
df  = var_quali.copy()
keys = df.keys()
for i in range(len(keys)):
    for j in range(len(keys)):
        column1 = keys[i]
        column2 = keys[j]
        if j>i:
            print(column1 + ' x ' + column2)
            contingencia_chi(df, column1, column2)
            print('\n')
#%%

# 02° Passo: Número de dimensões (número de categorias - número de variáveis)

# Primeiro encontrar a quantidade de dimensões, isto é a quantidade de categorias menos a quantidade de variáveis
numero_categorias = 0
for variable in df.keys():
    numero_categorias = numero_categorias + df[variable].nunique()
    
numero_variaveis = len(df.keys())

numero_dimensoes = numero_categorias - numero_variaveis

#%%

# 03° Passo: ANACOR pelo Prince

# Análise de correspondência múltipla feita pelo "prince". Os valores das coordenadas são diferentes dos encontrados no R.

mca = prince.MCA(n_components=numero_dimensoes, n_iter=3, copy=True, check_input=True,engine='auto', random_state=42)
mca = mca.fit(df)
#%%

# 04° Passo: Mapa perceptual

ax = mca.plot_coordinates(df, ax=None, figsize=(6, 6),show_row_points=False,
                             show_column_points=True, column_points_size=30, show_column_labels=True, legend_n_cols=2)

ax.legend(bbox_to_anchor=(1.3, 1.0))
ax.get_figure()

print(mca.row_coordinates(df))
print(mca.column_coordinates(df))


print(mca.eigenvalues_)
print(mca.total_inertia_)
print(mca.explained_inertia_)

#%%

# Adicionando as coordenadas ao banco de dados de variáveis quantitativas
coord_obs = mca.row_coordinates(df)
var_quanti['Axis1'] = coord_obs.iloc[:, 0]
var_quanti['Axis2'] = coord_obs.iloc[:, 1]


#%%

# FIM ANACOR PARA VARIÁVEIS QUALITATIVAS

#%%

# INÍCIO PCA PARA VARIÁVEIS QUANTITATIVAS

#%%

# 01º Passo: Matriz de correlações de Pearson

# Coeficientes de correlação de Pearson para cada par de variáveis
def pearson(df1, df2):

    coeffmat = np.zeros((df1.shape[1], df2.shape[1]))
    pvalmat = np.zeros((df1.shape[1], df2.shape[1]))
    
    for i in range(df1.shape[1]):    
        for j in range(df2.shape[1]):        
            corrtest = pearsonr(df1[df1.columns[i]], df2[df2.columns[j]])  
    
            coeffmat[i,j] = corrtest[0]
            pvalmat[i,j] = corrtest[1]

    corr_coef = pd.DataFrame(coeffmat, columns=df2.columns, index=df1.columns)
    corr_sig = pd.DataFrame(pvalmat, columns=df2.columns, index=df1.columns)
    return corr_coef, corr_sig


df =  var_quanti.copy()
corr_coef, corr_sig = pearson(df, df)
print('##### MATRIZ DE CORRELAÇÃO ######') #OK
print(corr_coef)
print('##### P-VALUE ######') #OK
print(corr_sig)

#%%
# Elaboração de um mapa de calor das correlações de Pearson entre as variáveis
ax = sns.heatmap(
    corr_coef, 
    vmin=-1, vmax=1, center=0,
    cmap=sns.diverging_palette(20, 220, n=200),
    square=True,
    annot=True, 
    fmt='.2f')

ax.set_xticklabels(
    ax.get_xticklabels(),
    rotation=45,
    horizontalalignment='right'
)

#%%

# 02° Passo: Teste de esferecidade de Barllet

def bartlett_sphericity(dataset, corr_method="pearson"):
    
    r"""
    
    Parameters
    ----------
    dataset : dataframe, mandatory (numerical or ordinal variables)
        
    corr_method : {'pearson', 'spearman'}, optional
        
    Returns
    -------
    out : namedtuple
        The function outputs the test value (chi2), the degrees of freedom (ddl)
        and the p-value.
        It also delivers the n_p_ratio if the number of instances (n) divided 
        by the numbers of variables (p) is more than 5. A warning might be issued.
        
        Ex:
        chi2:  410.27280642443156
        ddl:  45.0
        p-value:  8.73359410503e-61
        n_p_ratio:    20.00
        
        Out: Bartlett_Sphericity_Test_Results(chi2=410.27280642443156, ddl=45.0, pvalue=8.7335941050291506e-61)
    
    References
    ----------
    
    [1] Bartlett,  M.  S.,  (1951),  The  Effect  of  Standardization  on  a  chi  square  Approximation  in  Factor
    Analysis, Biometrika, 38, 337-344.
    [2] R. Sarmento and V. Costa, (2017)
    "Comparative Approaches to Using R and Python for Statistical Data Analysis", IGI-Global.
    
    Examples
    --------
    illustration how to use the function.
    
    >>> bartlett_sphericity(survey_data, corr_method="spearman")
    chi2:  410.27280642443145
    ddl:  45.0
    p-value:  8.73359410503e-61
    n_p_ratio:    20.00
    C:\Users\Rui Sarmento\Anaconda3\lib\site-packages\spyderlib\widgets\externalshell\start_ipython_kernel.py:75: 
    UserWarning: NOTE: we advise  to  use  this  test  only  if  the number of instances (n) divided by the number of variables (p) is lower than 5. Please try the KMO test, for example.
    backend_o = CONF.get('ipython_console', 'pylab/backend', 0)
    Out[12]: Bartlett_Sphericity_Test_Results(chi2=410.27280642443156, ddl=45.0, pvalue=8.7335941050291506e-61)
    """
    
    import numpy as np
    import math as math
    import scipy.stats as stats
#    import warnings as warnings
    import collections

    #Dimensions of the Dataset
    n = dataset.shape[0]
    p = dataset.shape[1]
#    n_p_ratio = n / p
    
    #Several Calculations
    chi2 = - (n - 1 - (2 * p + 5) / 6) * math.log(np.linalg.det(dataset.corr(method=corr_method)))
    #Freedom Degree
    ddl = p * (p - 1) / 2
    #p-value
    pvalue = 1 - stats.chi2.cdf(chi2 , ddl)
    
    Result = collections.namedtuple("Bartlett_Sphericity_Test_Results", ["chi2", "ddl", "pvalue"],  rename=False)   
    
    #Output of the results - named tuple
    result = Result(chi2=chi2,ddl=ddl,pvalue=pvalue) 

    
    #Output of the function
    # if n_p_ratio > 5 :
    #     print("n_p_ratio: {0:8.2f}".format(n_p_ratio))
    #     warnings.warn("NOTE: we advise  to  use  this  test  only  if  the number of instances (n) divided by the number of variables (p) is lower than 5. Please try the KMO test, for example.")
        
    
    return result

bartlett_chisq, bartlett_degrees, bartlett_pvalue = bartlett_sphericity(df, 'pearson')

print('##### CHI-SQUARED ######') #OK
print(bartlett_chisq)
print('##### DEGREES OF FREEDOM ######') #OK
print(bartlett_degrees)
print('##### P-VALUE ######') #OK
print(bartlett_pvalue)


#%%

# 03° Passo: Autovalores e autovetores da matriz de correlação de Pearson. Da pra tirar a variância compartilhada, que é a porcentagem de cada autovalor da soma deles

# Autovalores e autovetores
eigenvalues,eigenvectors = sla.eig(corr_coef)
eigenvalues = np.real(eigenvalues)

idx = eigenvalues.argsort()[::-1]   
eigenvalues = eigenvalues[idx]
eigenvectors = eigenvectors[:,idx]
eigenvectors[:, 1] = -eigenvectors[:, 1]
eigenvectors = -eigenvectors
#%%
print('##### AUTOVALORES ######') #OK
print(eigenvalues)
print('\n')
print('##### VARIÂNCIAS COMPARTILHADAS ######') #OK
eigenvalues_variance = eigenvalues/np.sum(eigenvalues)
print(eigenvalues_variance)
#%%
print('##### AUTOVETORES ######') #OK, sinais diferentes mas não deveria mudar os resultados
print(eigenvectors)

#%%

# 04° Passo: Scores fatoriais (autovetores dividido pela raiz do seu autovalor)

# Scores fatoriais 
fatoriais = eigenvectors/np.sqrt(eigenvalues[None,:]) #OK, sinais diferentes mas não deveria mudar os resultados
print(fatoriais)

#%%

# 05° Passo: Zscores (são as observações padronizadas)

# Z-scores
df_zscore = zscore(df, ddof = 1)
print('##### Z-SCORES ######') #OK
print(df_zscore)

#%%

# 06° Passo: Fatores (multiplicação matricial dos Zscores com os scores fatoriais)

# Fatores
fatores = np.matmul(df_zscore, fatoriais)
print('##################### FATORES ####################') #OK, sinais diferentes mas não deveria mudar os resultados
print(fatores)
print('\n')
print('################# PEARSON FATORES ################') #OK
print(pearson(fatores, fatores)[0])

#%%

# 07° Passo: Cargas fatoriais (correlaçaõ de Pearson entre os fatores e as variáveis originais)

cargas_fatoriais, _ = pearson( df, fatores)
print('################ CARGAS FATORIAIS ################') #OK, sinais diferentes mas não deveria mudar os resultados
print(cargas_fatoriais)
print('\n')
print('################# COMUNALIDADES ##################') #OK
print(np.sum(cargas_fatoriais**2, axis = 1))
print('\n')
print('################## AUTOVALORES ###################') #OK
print(np.sum(cargas_fatoriais**2, axis = 0))
print(eigenvalues)

#%%

# 08° Passo: Critério de Kaiser (autovalores maiores que 1), para definir o número de fatores
# 09° Passo: Comunalidades dos autovalores de Kaiser

# Critério de Kaiser
kaiser = np.where(eigenvalues >1)[0]
cargas_fatoriais_kaiser = cargas_fatoriais.iloc[:, kaiser.tolist()]
print('############# CARGAS FATORIAIS KAISER ############')
print(cargas_fatoriais_kaiser)
print('\n')
print('############## COMUNALIDADES KAISER ##############')
print(np.sum(cargas_fatoriais_kaiser**2, axis = 1))

#%%

# 10° Passo: Plotar as cargas fatoriais (loading plot)

# Loading plot
fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(1, 1, 1)
x =  cargas_fatoriais_kaiser.iloc[:,0]
y = cargas_fatoriais_kaiser.iloc[:,1]
ax.scatter(x, 
           y,
           color = 'black',
           s = 60)
ax.axhline(0, color = 'grey', alpha = 0.5)
ax.axvline(0, color = 'grey', alpha = 0.5)


ax.set_xlabel('Principal Component 1',
              fontsize = 20)
ax.set_ylabel('Principal Component 2',
              fontsize = 20)

ax.set_xlim(-1.1, 1.1)
ax.set_ylim(-1.1, 1.1)

text =  cargas_fatoriais_kaiser.index.values
for i in range(len(x)):
    plt.annotate(text[i], (x[i], y[i] + 0.02))
  

plt.show()
#%%
# Adicionando os fatores extraídos no banco de dados original
for i in kaiser:
    column_name = 'fator_' + str(i + 1)
    df[column_name] = fatores.iloc[:,i].values


# FIM DA ANÁLISE POR COMPONENTES PRINCIPAIS    
        

